### ISSUE TITLE
How to Implement Early Stopping with WebDataset in SageMaker Distributed Data Parallel (SMDDP) Framework?

### ISSUE THREAD

**Issue by nov05** [webdataset/webdataset#446]
* I am currently using `WebDataset` with the AWS `SMDDP framework` (backend='smddp') for distributed training, and everything works fine when training runs for the specified number of epochs. However, when I try to implement an 'early stopping' mechanism—such as breaking the epoch loop if the training loss doesn't improve for 5 epochs—I encounter an issue. Specifically, the broadcasting process stops as expected, but one or more other processes hang until the SageMaker timeout error is triggered.

* Below is the code that implements early stopping using `dist.broadcast()` and `dist.barrier()`.  

```python
import smdistributed.dataparallel.torch.distributed as dist
from smdistributed.dataparallel.torch.parallel.distributed import DistributedDataParallel as DDP
dist.init_process_group(backend="smddp")  ## SageMaker DDP, replacing "nccl"

def main(task):
    ...
    task.model = torch.nn.SyncBatchNorm.convert_sync_batchnorm(task.model)  ## PyTorch DDP
    ## Wrap model with SMDDP
    task.model = DDP(task.model.to(task.config.device), 
                     device_ids=[torch.cuda.current_device()])  ## single-device Torch module 
    ...
    # braodcast_early_stop = torch.tensor(0, dtype=torch.int32).to(task.config.device)
    for epoch in range(task.config.epochs):
        task.current_epoch = epoch
        if dist.get_rank()==0:
            print(f"👉 Train Epoch: {epoch}, "
                  f"Learning Rate: {task.optimizer.param_groups[0]['lr']}")
## ❌ Some processes get stuck here after the broadcasting process (rank=0) printing out the above message 
## if the 'early stopping' code is uncommented.
       
        train(task)

        if dist.get_rank()==0:
            eval(task, phase='val')
            # if task.early_stopping.early_stop: 
            #     ## broadcast the early stop decision to all nodes
            #     braodcast_early_stop = torch.tensor(1, 
            #         dtype=torch.int32).to(task.config.device)
            #     dist.broadcast(braodcast_early_stop, src=0)  ## one to all, src is the process rank
            #     print(f"⚠️ Early stopping broadcasting "
            #         f"{braodcast_early_stop.item()} from Rank {dist.get_rank()}...")

            ## adjust optimizer learning rate
            if task.config.opt_type=='adamw':
                task.scheduler.step()  
            elif task.config.opt_type=='sgd':
                adjust_learning_rate(
                    task.config.opt_learning_rate, 
                    task.config.lr_sched_step_size,
                    task.current_epoch+1
                )

        # if braodcast_early_stop.item()!=0:
        #     print(f"⚠️ Early stopping at epoch {task.current_epoch} "
        #         f"on node {dist.get_rank()}")
        #     break
## ❌ The above message only gets printed once by the broadcasting process (rank=0).

    if dist.get_rank()==0:
        print("🟢 Start testing...")
        eval(task, phase='test')
        print("🟢 Start saving the trained model...")
        save(task)

if __name__="__main__":
    ...
    main()
    ...

```

* I have gone through some posts here but still don't know whether it's possible to implement 'early stopping' or if there is a workaround solution. I would be grateful for any advice. Thank you!

[tmbdev](https://github.com/tmbdev) [on Jul 24, 2021](https://github.com/webdataset/webdataset/issues/90#issuecomment-886086050) Collaborator
> WebDataset doesn't kill processes at all, it just creates a Pipe object and runs whatever command you ask it to in it. The webdataset reader is finished whenever the job finishes, so there is usually no need to kill anything. The only way processes hang around is if you start an epoch, stop before it finishes, and then start another one.

**Comment by tmbdev** [comment]
I'm not sure why it is hanging or why this would work with other backends. Unless you use it with num_workers=0, the DataLoader creates worker processes and they read. When you stop early, I think the DataLoader just creates a new set of workers for the next epoch, unless you ask it to do otherwise. The old worker processes don't even share file descriptors with the DataLoader process or the new worker processes, so it's not like you are going to run out of file descriptors either. Because workers are subprocesses, the file descriptors just get closed when the worker process gets destroyed. If the worker process doesn't get destroyed, you wouldn't notice for a long time until maybe eventually you run out of kernel resources.

Because the worker process is separate from your training process, you couldn't even close the file descriptors if you tried; all you can do is kill the subprocesses that the DataLoader created.

So, I don't know what the problem could be.

To diagnose this, try to get a process tree that shows what is actually going on with the process, and then possibly attach to the processes with the Python debugger to see where it is hanging.

**Comment by nov05** [comment]
> I'm not sure why it is hanging or why this would work with other backends. Unless you use it with num_workers=0, the DataLoader creates worker processes and they read. When you stop early, I think the DataLoader just creates a new set of workers for the next epoch, unless you ask it to do otherwise. The old worker processes don't even share file descriptors with the DataLoader process or the new worker processes, so it's not like you are going to run out of file descriptors either. Because workers are subprocesses, the file descriptors just get closed when the worker process gets destroyed. If the worker process doesn't get destroyed, you wouldn't notice for a long time until maybe eventually you run out of kernel resources.
> 
> Because the worker process is separate from your training process, you couldn't even close the file descriptors if you tried; all you can do is kill the subprocesses that the DataLoader created.
> 
> So, I don't know what the problem could be.
> 
> To diagnose this, try to get a process tree that shows what is actually going on with the process, and then possibly attach to the processes with the Python debugger to see where it is hanging.

Thank you so much for the helpful information! 🙏 I can now confirm that the issue was related to how to properly breaking the loop with the SageMaker-Torch DDP framework rather than WebDataset. I simply replaced `dist.broadcast(broadcast_early_stop, src=0)` with `dist.all_reduce(broadcast_early_stop, op=dist.ReduceOp.SUM)`, and the error is gone.
